lib/logic/player/favorites.dart

```diff
@@ -1,27 +1,193 @@
+    Settings.useMediaStoreForFavoriteSongs.value;
+
+  MediaStoreContentObserver<Song>? _mediaStoreContentObserver;
+
+  @override
+  Future<void> init() async {
+    super.init();
+    _showOnlyFavoritesNotifier.value = false;
+    for (final contentType in Content.enumerate()) {
+      await _initContentType(contentType);
+    }
+    _registerMediaStoreObserver();
+    Settings.useMediaStoreForFavoriteSongs.addListener(_useMediaStoreForFavoriteSongsListener);
+  }
+
+  @override
+  void dispose() {
+    _favoriteSetsMap.clear();
+    _disposeMediaStoreObserver();
+    Settings.useMediaStoreForFavoriteSongs.removeListener(_useMediaStoreForFavoriteSongsListener);
+    super.dispose();
+  }
+
+  Future<void> _initContentType(Type contentType) async {
+    final Set<int> favoriteSet = {};
+    if (_useMediaStoreFavorites(contentType)) {
+      for (final song in ContentControl.instance.state.allSongs.songs) {
+        if (song.isFavoriteInMediaStore!) {
+          favoriteSet.add(song.sourceId);
+        }
+      }
+    } else {
+      final serializer = repository.serializersMap.getValue(contentType);
+      await serializer!.init();
+      final savedIds = await serializer.read();
+      favoriteSet.addAll(savedIds);
+    }
+    _favoriteSetsMap.setValue(favoriteSet, key: contentType);
+  }
+
+  Future<void> _useMediaStoreForFavoriteSongsListener() async {
+    if (Settings.useMediaStoreForFavoriteSongs.value) {
+      _registerMediaStoreObserver();
+    } else {
+      _disposeMediaStoreObserver();
+    }
+    await _initContentType(Song);
+    ContentControl.instance.emitContentChange();
+  }
+
+  void _registerMediaStoreObserver() {
+    _mediaStoreContentObserver = MediaStoreContentObserver()
+      ..onChangeStream.listen(_handleMediaStoreSongsChange)
+      ..register();
+  }
+  void _disposeMediaStoreObserver() {
+    _endMediaStoreUpdate();
+    _mediaStoreContentObserver?.dispose();
+    _mediaStoreContentObserver = null;
+  }
+
+  static const _mediaStoreUpdateDebounceInterval = Duration(milliseconds: 200);
+  Timer? _mediaStoreUpdateTimer;
+  void _endMediaStoreUpdate() {
+    _mediaStoreUpdateTimer?.cancel();
+    _mediaStoreUpdateTimer = null;
+  }
+  void _handleMediaStoreSongsChange(MediaStoreContentChangeEvent event) {
+    if (event.flags & AndroidContentResolver.NOTIFY_UPDATE != 0) {
+      _mediaStoreUpdateTimer?.cancel();
+      _mediaStoreUpdateTimer = Timer(_mediaStoreUpdateDebounceInterval, () async {
+        await ContentControl.instance.refetch<Song>();
+        final favoriteSet = _favoriteSetsMap.getValue<Song>()!;
+        for (final song in ContentControl.instance.state.allSongs.songs) {
+          if (song.isFavoriteInMediaStore! && !favoriteSet.contains(song.id)) {
+            favoriteSet.add(song.id);
+          } else if (!song.isFavoriteInMediaStore! && favoriteSet.contains(song.id)) {
+            favoriteSet.remove(song.id);
+          }
+        }
+        _endMediaStoreUpdate();
```

Comment:
Moved it to the top.

> https://developer.android.com/reference/kotlin/android/database/ContentObserver

We don't need to refresh until updates stop coming, because it is a relatively expensive operation.